pkg/cli/cliexecuter.go

```diff
@@ -1,97 +1,236 @@
 
-// NewExecuter builds the executer
-func NewExecuter(config *restclient.Config) (Executer, error) {
+// NewExecuter builds the Executer
+func NewExecuter(client client.Client) (*Executer, error) {
 	logger := ctrl.Log.WithName("executer")
 
-	clientSet, err := kubernetes.NewForConfig(config)
-	if err != nil {
-		logger.Error(err, "failed building k8s client")
-		return nil, err
-	}
+	return &Executer{
+		Client:   client,
+		log:      logger,
+		routines: make(map[types.UID]*routine),
+		runner:   run,
+	}, nil
+}
 
-	return &executer{
-		log:       logger,
-		config:    config,
-		clientSet: clientSet,
+// NewExecuter builds an Executer with configurable runnerFunc for testing
+func NewFakeExecuter(client client.Client, fn runnerFunc) (*Executer, error) {
+	logger := ctrl.Log.WithName("fakeExecuter")
+
+	return &Executer{
+		Client:   client,
+		log:      logger,
+		routines: make(map[types.UID]*routine),
+		runner:   fn,
 	}, nil
 }
 
-// Execute builds and runs a Post request on contianer for SPDY (shell) executor
-func (e executer) Execute(pod *corev1.Pod, command []string) (stdout string, stderr string, err error) {
-	if len(pod.Spec.Containers) == 0 {
-		err := errors.New("create cli executer failed")
-		e.log.Error(err, "No container found in Pod", "Pod Name", pod.Name)
-		return "", "", err
+// AsyncExecute runs the command in a goroutine mapped to the UID
+func (e *Executer) AsyncExecute(ctx context.Context, uid types.UID, command []string, retryCount int, retryDuration, timeout time.Duration) {
+	e.mapLock.Lock()
+	defer e.mapLock.Unlock()
+	if _, exist := e.routines[uid]; !exist {
+		// create a context for the fence agent command that the controller can cancel
+		cancellableCtx, cancel := context.WithCancel(ctx)
+		routine := routine{
+			cancel: cancel,
+		}
+		e.routines[uid] = &routine
+
+		go func(faCtx, crCtx context.Context, uid types.UID, command []string) {
+			// Linear backoff
+			backoff := wait.Backoff{
+				Steps:    retryCount,
+				Duration: retryDuration,
+				Factor:   1.0,
+			}
+
+			var stdout, stderr string
+			var faErr error
+
+			e.log.Info("fence agent start", "uid", uid, "command", command, "retryCount", retryCount, "retryDuration", retryDuration, "timeout", timeout)
+			// ExponentialBackoff to handle the fence agent command execution where:
+			// - the command fails: the command is retried until the retryCount is reached
+			// - the command times out: the command is retried until the retryCount is reached
+			// - the FA context times out: the command is cancelled and the status is updated
+			// - the FA context is cancelled: the command is cancelled and the status is not updated
+			// - the command succeeds: the command is not retried and the status is updated
+			err := wait.ExponentialBackoffWithContext(faCtx,
+				backoff,
+				func(ctx context.Context) (bool, error) {
+					ctxWithTimeout, cancel := context.WithTimeout(faCtx, timeout)
```

Comment:
what about extracting this into its own func? Same for the second ExponentialBackoffWithContext call

Reply:
The updated version is different from what I meant, but it's better than the old version a well, so I'm fine with it 🙂 
( I meant to leave wait.ExponentialBackoffWithContext() here, and just extract the function argument into a separate function)